src/libcollections/linked_list.rs

```diff
@@ -35,222 +30,156 @@ use super::SpecExtend;
 /// A doubly-linked list.
 #[stable(feature = "rust1", since = "1.0.0")]
 pub struct LinkedList<T> {
-    length: usize,
-    list_head: Link<T>,
-    list_tail: Rawlink<Node<T>>,
-}
-
-type Link<T> = Option<Box<Node<T>>>;
-
-struct Rawlink<T> {
-    p: Option<Shared<T>>,
+    head: Option<Shared<Node<T>>>,
+    tail: Option<Shared<Node<T>>>,
+    len: usize,
+    marker: PhantomData<Box<Node<T>>>,
```

Comment:
Not strictly needed since `Shared<X>` contains `PhantomData<X>` and `Box<X>` and `X` have the same implications.